### PR TITLE
Add ability to set Authorization header Scheme

### PR DESCRIPTION
**Pull request by rigrassm**
Took a quick pass at implementing this to address issue #306.

This PR adds a new Client/Request method named `SetAuthScheme` that allows the user to specify an Authorization header scheme. If one isn't set, the `addCredentials()` function will see this and default to using the "Bearer" scheme.

I'm a new Gopher so please be critical of the work!

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -294,6 +295,22 @@ func (c *Client) SetAuthToken(token string) *Client {
 	return c
 }
 
+// SetAuthScheme method sets the auth token scheme type in the HTTP request. For Example:
```
Comment (jeevatkm) [inline client.go]: @rigrassm I think its good to include the RFC and IANA links with few texts in the godoc would be helpful for the user.

File: request.go [hunk 2]
```diff
@@ -395,6 +396,22 @@ func (r *Request) SetAuthToken(token string) *Request {
 	return r
 }
 
+// SetAuthScheme method sets the auth token scheme type in the HTTP request. For Example:
```
Comment (jeevatkm) [inline request.go]: @rigrassm I think its good to include the RFC and IANA links with few texts in the godoc would be helpful for the user.

**Comment by jeevatkm** [comment]
@rigrassm ping 😄

**Comment by rigrassm** [comment]
Sorry for the delay, getting to the end of the quarter at work so things are a bit busy. I'll try to sneak in some time to knock this out in the best day or two.

File: request.go (hunk 2 above)
Comment (rigrassm) [reply request.go]: Added in both client.go and request.go

File: client.go [hunk 3]
```diff
@@ -294,6 +295,27 @@ func (c *Client) SetAuthToken(token string) *Client {
...
```
Comment (moorereason) [inline client.go]: s/SetAuthScheme/The SetAuthScheme/
s/token//

I'd also reword this to be like my wordsmithing suggestion on SetAuthToken.

File: client.go [hunk 4]
```diff
@@ -294,6 +295,27 @@ func (c *Client) SetAuthToken(token string) *Client {
...
+//      Authorization: <auth-scheme-value-set-here> <auth-token-value>
+//
+// For Example: To set the scheme to use OAuth
+//
+// 		client.SetAuthScheme("OAuth")
+//
+// This auth token gets added to all the request rasied from this client instance.
+// Also it can be overridden or set one at the request level is supported.
+//
+// Information about Auth schemes can be found in RFC7235 which is linked to below along with the page containing
```
Comment (moorereason) [inline client.go]: You capitalize "Auth" here, don't in most other places, and spell it out in the next line.  Can we be more consistent?  I would personally just spell out the words each time.  It's the authentication scheme on the Authorization header.

File: client.go [hunk 5]
```diff
@@ -294,6 +295,27 @@ func (c *Client) SetAuthToken(token string) *Client {
...
 
+// SetAuthScheme method sets the auth token scheme type in the HTTP request. For Example:
+//      Authorization: <auth-scheme-value-set-here> <auth-token-value>
+//
+// For Example: To set the scheme to use OAuth
+//
+// 		client.SetAuthScheme("OAuth")
+//
+// This auth token gets added to all the request rasied from this client instance.
+// Also it can be overridden or set one at the request level is supported.
```
Comment (moorereason) [inline client.go]: It sounds better to my ear to say:

> This authentication scheme gets added to all requests raised from this client instance but can be overridden at the request level.

File: client.go [hunk 6]
```diff
@@ -278,14 +279,14 @@ func (c *Client) SetBasicAuth(username, password string) *Client {
 	return c
 }
 
-// SetAuthToken method sets bearer auth token header in the HTTP request. For Example:
-// 		Authorization: Bearer <auth-token-value-comes-here>
+// SetAuthToken method sets the auth token header (default scheme: Bearer) in the HTTP request. For Example:
```
Comment (moorereason) [inline client.go]: Wordsmithing...

> The SetAuthToken method sets the authentication token of the Authorization header for all HTTP requests.  The default authentication scheme is Bearer, but this can be changed with SetAuthScheme.

File: client.go [hunk 7]
```diff
@@ -278,14 +279,14 @@ func (c *Client) SetBasicAuth(username, password string) *Client {
...
+// 		Authorization: <auth-scheme> <auth-token-value-comes-here>
```
Comment (moorereason) [inline client.go]: s/auth-token-value-comes-here/auth-token/

File: client.go [hunk 8]
```diff
@@ -294,6 +295,27 @@ func (c *Client) SetAuthToken(token string) *Client {
 	return c
 }
 
+// SetAuthScheme method sets the auth token scheme type in the HTTP request. For Example:
+//      Authorization: <auth-scheme-value-set-here> <auth-token-value>
```
Comment (moorereason) [inline client.go]: s/-value-set-here//
s/-value//

File: middleware.go [hunk 9]
```diff
@@ -252,11 +252,21 @@ func addCredentials(c *Client, r *Request) error {
...
+	} else if !IsStringEmpty(c.AuthScheme) {
+		authScheme = c.AuthScheme
+	} else {
+		authScheme = "Bearer"
+	}
+
+	// Build the Token Auth header
 	if !IsStringEmpty(r.Token) { // takes precedence
-		r.RawRequest.Header.Set(hdrAuthorizationKey, "Bearer "+r.Token)
+		r.RawRequest.Header.Set(hdrAuthorizationKey, fmt.Sprintf("%s %s", authScheme, r.Token))
```
Comment (moorereason) [inline middleware.go]: I suspect that simple string concatenation will be more efficient.

**Comment by jeevatkm** [comment]
hello @rigrassm, @moorereason have shared a few inputs and comments. Could you please take care?

File: client.go [hunk 10]
```diff
@@ -278,14 +279,14 @@ func (c *Client) SetBasicAuth(username, password string) *Client {
...
-// 		Authorization: Bearer <auth-token-value-comes-here>
+// SetAuthToken method sets the auth token header (default scheme: Bearer) in the HTTP request. For Example:
+// 		Authorization: <auth-scheme> <auth-token-value-comes-here>
 //
 // For Example: To set auth token BC594900518B4F7EAC75BD37F019E08FBC594900518B4F7EAC75BD37F019E08F
 //
 // 		client.SetAuthToken("BC594900518B4F7EAC75BD37F019E08FBC594900518B4F7EAC75BD37F019E08F")
 //
-// This bearer auth token gets added to all the request rasied from this client instance.
+// This auth token gets added to all the request rasied from this client instance.
```
[the remark this answers is not in the document]
Comment (jeevatkm) [reply client.go]: Done in my branch will push it to rigrassm:master

File: client.go [hunk 11]
```diff
@@ -294,6 +295,27 @@ func (c *Client) SetAuthToken(token string) *Client {
...
 }
 
+// SetAuthScheme method sets the auth token scheme type in the HTTP request. For Example:
+//      Authorization: <auth-scheme-value-set-here> <auth-token-value>
+//
+// For Example: To set the scheme to use OAuth
+//
+// 		client.SetAuthScheme("OAuth")
+//
+// This auth token gets added to all the request rasied from this client instance.
```
[the remark this answers is not in the document]
Comment (jeevatkm) [reply client.go]: Done in my branch will push it to rigrassm:master

File: client.go (hunk 3 above)
Comment (jeevatkm) [reply client.go]: Method name has to start first and the `token` part taken care in my branch will push it to rigrassm:master

File: client.go (hunk 4 above)
Comment (jeevatkm) [reply client.go]: Done in my branch will push it to rigrassm:master

File: client.go (hunk 6 above)
Comment (jeevatkm) [reply client.go]: SetAuthToken method sets the auth token of the `Authorization` header for all HTTP requests. The default auth scheme is `Bearer`, it can be customized with the method `SetAuthScheme`.

File: client.go (hunk 8 above)
Comment (jeevatkm) [reply client.go]: Done in my branch will push it to rigrassm:master

File: middleware.go (hunk 9 above)
Comment (jeevatkm) [reply middleware.go]: This [gist page](https://gist.github.com/dtjm/c6ebc86abe7515c988ec) has some data on this topic. It seems `Buffer`, `+`, and so on.  I will update it to `+`. Thanks.  

Done in my branch will push it to rigrassm:master
